python/runtime/cudaq/platform/py_alt_launch_kernel.cpp

```diff
@@ -634,10 +634,46 @@ std::string getASM(const std::string &name, MlirModule module,
   auto cloned = unwrap(module).clone();
   auto context = cloned.getContext();
 
+  // Get additional debug values
+  auto disableMLIRthreading = getEnvBool("CUDAQ_MLIR_DISABLE_THREADING", false);
+  auto enablePrintMLIREachPass =
+      getEnvBool("CUDAQ_MLIR_PRINT_EACH_PASS", false);
+
   PassManager pm(context);
   pm.addPass(cudaq::opt::createLambdaLiftingPass());
+  // Run most of the passes from hardware pipelines.
+
+  pm.addPass(createCanonicalizerPass());
+  pm.addPass(createCSEPass());
+  pm.addNestedPass<func::FuncOp>(cudaq::opt::createClassicalMemToReg());
+  pm.addPass(cudaq::opt::createLoopUnroll());
+  pm.addPass(createCanonicalizerPass());
```

Comment:
Make these two Func level so as to get better locality and performance here.

Reply:
There is no reason canonicalize can't be run with a nested pass. Something strange is going on.

Reply:
it can, I meant the loop unroll. I moved everything that can run as a func pass to func level.

Reply:
will update after the #2422 lands